typescript/introduction/tsc-variable-declaration/tsc-block-scoping.md

```diff
@@ -0,0 +1,98 @@
+---
+author: kapnobatai136
+
+type: normal
+
+category: must-know
+
+aspects:
+  - introduction
+
+---
+
+# Block-Scoping
+
+---
+## Content
+
+Let's talk a bit more about how the block scope works.
+
+We know that when a variable is defined using `let` or `const`, it's only accessible in the containing pair of curly braces, i.e. `{}`.
+
+Let's take another look at an example to understand this better:
+
+```ts
+function theBestApp() {
+  let getApp;
+
+  if (true) {
+    let app = 'Enki';
+    getApp = function() {
+      return app;
+    }
+  }
+
+  return getApp()
+}
+console.log(theBestApp());
+// Enki
+```
+
+Because we first define `getApp` and then we assign it a function, we won't get any errors.
+
+It should be noted that every iteration of a loop creates a new block-scope.
+
+```ts
+// the variable i is redeclared
+// in each loop iteration
```

Comment:
This is not exactly true. In each iteration, the value of `i` is **redefined** and since the value is primitive and because of scope of i being inside the loop, the function in setTimeout refers back to the value of `i` in that iteration. With a reference type you can see that its only reassignment in 3rd part of loop being triggered, and no redeclaration. In case of `var` the scope is moved up a level and even though the value is primitive and is reassigned, all the timeout calls are referring to same variable.

```js
for (let i = 0, obj={a: 0}; i < 3; i++, obj.a=i) {
  setTimeout(function() { console.log(obj, i) }, 1000 * i)
};

// {a: 3} 0
// {a: 3} 1
// {a: 3} 2
```

Reply:
Nice catch.

@kapnobatai137 let's say

```ts
// each loop iteration
// defines a new i variable
```

Reply:
Isn't that saying the same thing? It should instead be
```
// each loop iteration
// assigns a new value to i variable
```

Reply:
That's not the case though, right? Each loop iteration does have a new variable (because each loop iteration creates a new block scope).

Each loop iteration is its own new scope instance.

Reply:
So re-declaration is wrong not because nothing is being declare multiple times but because it's an entirely new declaration each time.

it's like this:

```js
// loop 1
{
  let i = 0;
  // do stuff
}

// loop 2
{
  let i = 1;
  // do stuff
}
```

So it's multiple declarations rather than a re-declaration

```js
// this is re-declaration, which isn't happening
let i = 0;
let i = 1;
```